definitions/scenarios/self_upgrade.rb

```diff
@@ -58,15 +66,38 @@ class SelfUpgrade < SelfUpgradeBase
       manual_detection
     end
 
-    def compose
+    def downstream_self_upgrade(pkgs_to_update)
       if check_min_version('foreman', '2.5') || check_min_version('foreman-proxy', '2.5')
-        pkgs_to_update = %w[satellite-maintain rubygem-foreman_maintain]
         yum_options = req_repos_to_update_pkgs.map do |id|
           "--enablerepo=#{id}"
         end
         add_step(Procedures::Packages::Update.new(packages: pkgs_to_update, assumeyes: true,
                                                   yum_options: yum_options))
       end
     end
+
+    def upstream_self_upgrade(pkgs_to_update)
+      add_step(Procedures::Repositories::Setup.new(:version => upstream_target_version))
+      add_step(Procedures::Packages::Update.new(packages: pkgs_to_update, assumeyes: true))
+      installed_release_pkg = package_manager.find_installed_package('foreman-release',
+                                                                     '%{VERSION}')
+      current_major_version = current_version[0..2]
+      installed_foreman_release_major_version = installed_release_pkg[0..2]
+      if installed_foreman_release_major_version != current_major_version
+        add_step(Procedures::Packages::Uninstall.new(packages: %w[foreman-release katello-repos],
+                                                     assumeyes: true))
+        add_step(Procedures::Repositories::Setup.new(:version => current_major_version))
+      end
```

Comment:
I had to look multiple times at this to understand what it's doing.
1. setup `target_version` repositories
2. update `foreman_maintain`
3. roll back the repositories to `current_version`

Given there is no good way to use a temporary repo definition (there is with DNF, but it would require us to get the `.repo` file from somewhere, which is also hard), I think this is as nice as it gets.

No idea if this deserves a comment or anything, but thought I mention that I stumbled on reading it.

Reply:
Added the comments and added `ensure` as well to rollback repositories.